### PR TITLE
Fix S3897: Classes that provide "Equals(<T>)" should implement "IEquatable<T>"

### LINKED ISSUE #462: Fix S3897: Classes that provide "Equals(<T>)" should implement "IEquatable<T>"
[RSPEC-3897](https://jira.sonarsource.com/browse/RSPEC-3897)

Reduce scope of the rule and create a new one [RSPEC-4035](https://jira.sonarsource.com/browse/RSPEC-4035)

### PR DESCRIPTION
**Pull request by Evangelink**
Fix #462

### CODE REVIEW

**Comment by Evangelink** [comment]
@michalb-sonar Don't review yet, we need to merge #463 first.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/EquatableRuleBase.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+
+using Microsoft.CodeAnalysis;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    public abstract class EquatableRuleBase : SonarDiagnosticAnalyzer
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/EquatableRuleBase.cs]: I don't think it is a good pattern to create a new base class just to share a couple of methods. IMHO, it adds too much complexity to be worth it. Also, if we want to reuse the methods in another class, it'd need to inherit from the class.

Wouldn't it be better if we moved IsValidEqualsMethodSymbol to a helper class? Then IsValidEquatableInterfaceSymbol and EqualsMethodName could be either shared (if suitable) or turned into a private method duplicated in both files. Duplication is bad, but for one-liners it's fine.

WDYT?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ClassWithEqualityShouldImplementIEquatable.cs [hunk 2]
```diff
@@ -55,94 +48,21 @@ protected sealed override void Initialize(SonarAnalysisContext context)
...
-
-                    if (HasNoTriggeringEqualsMethod(objectEqualsMethod, typeSpecificEqualsMethod) ||
-                        IsEquatableInterfaceCorrectlyImplemented(objectEqualsMethod, typeSpecificEqualsMethod, equatableInterface))
-                    {
-                        return;
-                    }
-
-                    c.ReportDiagnostic(CreateReportDiagnostic(equatableInterface != null, objectEqualsMethod,
-                        typeSpecificEqualsMethod, classDeclaration.Identifier));
+                    classSymbol.GetMembers(EqualsMethodName)
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ClassWithEqualityShouldImplementIEquatable.cs]: Perhaps invert if above and put this inside?

Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ClassWithEqualityShouldImplementIEquatable.cs]: This test is certainly a bit simpler and more readable than in was :)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/EquatableRuleBase.cs (hunk 1 above)
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/EquatableRuleBase.cs]: Agreed

**Pull request merged**